Contracts-v1

vanshwassan opened the pull request:
- Added Contracts
   - `WETH.sol`
   - `Payments.sol`
- Added scripts to:
   - Deploy WETH and Payments Contract
   - Set dAPI name
   - Send Payments to the contract
   - Withdraw payments from the contract (`onlyOwner`)
   - Check Payment Receipt (PRT) with `tokenId`

vanshwassan committed:
Added contracts

vanshwassan committed:
Added scripts

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,87 @@
+        );
+    }
+
+    function int224ToUint256(address token)
```
this function is probably not needed, you should be able to typecast it without this

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,87 @@
+
+    }
+
+    function getEthValue(address token)
```
I'm not sure what this function is doing

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,87 @@
+        view
+        returns (uint256)
+    {
+        uint256 newTokenPrice = (int224ToUint256(token));
```
again this is probably not needed

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,87 @@
+    function Payment(address token, uint256 _tokenAmount) public returns(uint256) {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        uint256 _usdValue = (getEthValue(token) * _tokenAmount)/10**18;
```
you should probably getting the decimals and not assume 10**18

vanshwassan committed:
fixed all the changes

vanshwassan committed:
Requested changes done

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,71 @@
+    mapping(uint256 => uint256) public TokenIDtoPrice; 
+
+constructor(address _dapiServer, address _ERC20Address) DapiReader(_dapiServer) ERC721("Payment Receipt", "PRT") {
+        _ERC20 = ERC20(_ERC20Address);
```
your fixing the address of the token in the constructor which means you can only do the payment with one token

vanshwassan replied:
Done, fixed

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,71 @@
+        int224 value = IDapiServer(dapiServer).readDataFeedValueWithDapiName(
+                DapiName
+        );
+        uint224 tokenPriceUint224 = uint224(value);
```
if the value is already int224 why are you again typecasting it to int224

vanshwassan replied:
typecasting it to `uint224` first

vanshwassan replied:
> The task is pretty simple, you need to
> 
>     1. Accept payments in multilple types of token
> 
>     2. Get their USD value
> 
>     3. Mint an NFT with the USD value
> 
> 
> The tests also aren't done properly because they are actually not testing anything (payment is made? but your not even checking if the `Dapiserver` is converting from the token value to usd value). Rewrite the contract and please pay attention to what is required and test it out yourself with the intended usage. The contract is not according to specs.

- Fixed all the issues, contract can now accept multiple ERC20 token payments. Added a function to add allowed tokens to be added. 

- Added another token, Mock Wrapped Matic.

- `ownerWithdrawFunds` will now withdraw all `allowedTokens` held by the smart contract. 

- `getTotalPayments` will now return total USD balance of all the allowed ERC20 tokens held by the smart contract.

- Added proper comments for each method.

- Didn't get the part where you said the `Payment` function isn't checking if the value is converted by calling the dAPI. The function will just throw an error: `Sender cannot read` if it's not able to read the dAPI. Also, added more error handling for it to be safer.

Ashar2shahid replied:
~~its already a uint224~~

Ashar2shahid replied:
```tokenPriceUint256 = uint256(uint224(value));```

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,71 @@
+    }
+    
+    // Function to get the total payments
+    function getTotalPayments() view public returns(uint) {
```
this only gets the total payments of once type of token

vanshwassan replied:
Fixed

Ashar2shahid commented:
The task is pretty simple, you need to 

1. Accept payments in multilple types of token
2. Get their USD value
3. Mint an NFT with the USD value

The tests also aren't done properly because they are actually not testing anything (payment is made? but your not even checking if the `Dapiserver` is converting from the token value to usd value). Rewrite the contract and please pay attention to what is required and test it out yourself with the intended usage. The contract is not according to specs.

vanshwassan committed:
Contract can accept multiple ERC20 token payments

vanshwassan commented:
@Ashar2shahid 
- Fixed all the issues, contract can now accept multiple ERC20 token payments. Added a function to add allowed tokens to be added.
- Added another token, Mock Wrapped Matic.
- `ownerWithdrawFunds` will now withdraw all `allowedTokens` held by the smart contract.
- `getTotalPayments` will now return total USD balance of all the allowed ERC20 tokens held by the smart contract.
- Added proper comments for each method.
- Didn't get the part where you said the Payment function isn't checking if the value is converted by calling the dAPI. The function will just throw an error: `Sender cannot read` if it's not able to read the dAPI. Also, added more error handling for it to be safer.

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,96 @@
+    ERC20 public _ERC20;
+    mapping(address => bytes32) public tokenDapiMapping;
+    mapping(uint256 => uint256) public TokenIDtoPrice; 
+    address[] allowedTokens;
```
its better to make this a mapping that maps address to a boolean, because arrays generally cost more gas

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,96 @@
+    }
+    // @dev add allowed tokens
+    function addAllowedTokens(address token) public onlyOwner {
+        allowedTokens.push(token);
```
again, use a mapping

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,96 @@
+        allowedTokens.push(token);
+    }
+    // @dev check if token is allowed
+    function tokenIsAllowed(address token) public view returns (bool) {
```
this function is not needed if you use a mapping

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,96 @@
+    }
+    
+    // @dev get all payments in USD
+    function getTotalPayments() public view returns(uint256) {
```
1. This should be parsing the tokenIds which already has the USD value at the time of the payment but that might still be a bad idea
2. This gets the price when you call the function not at the time of payment
3. There isn't a clean way to do this via a mapping, maybe call it `getContractBalance(address token)` and get the USD balance of the contract for a specific token

vanshwassan replied:
- Yeah but using a mapping would make `getTotalPayments()` and `ownerWithdrawFunds()` unusable for users if they have to add the address of each token manually to check/withdraw.

- Can just rename `getTotalPayments()` to `getContractBalance()`?

Ashar2shahid commented on contracts/Payments.sol:
```diff
@@ -0,0 +1,96 @@
+        return balance;
+    }
+    // @dev onlyOwner withdraw all tokens
+    function ownerWithdrawFunds() public onlyOwner {
```
probably change this to `ownerWithdrawFunds(address token)` if your changing it a mapping

vanshwassan replied:
That's the reason why I used an array instead of a mapping. Would have to loop through the ERC20 token addresses to transfer funds to the owner.

Ashar2shahid commented on deploy/3_deploy_Payments.js:
```diff
@@ -0,0 +1,66 @@
+  // Add allowed tokens
+  const addWETH = await paymentsContract.addAllowedTokens(WETHAddress);
+  await addWETH.wait();
+  console.log(`Wrapped ETH added to allowed tokens!`);
```
this isn't how you do tests, you need to test if the state of the contract has changed. refer to this: https://hardhat.org/tutorial/testing-contracts and https://www.chainshot.com/article/how-to-write-unit-tests and maybe see other DAO contracts later to see the pattern we follow

vanshwassan committed:
Added natspec comments: @vanshwassan

vanshwassan committed:
added licences to the token contracts

vanshwassan committed:
Update to use dapi proxy and new market contract.
- Updated contracts
- Updated deploy scripts

vanshwassan committed:
Changed function name
